src/renderer/components/cluster-manager/lens-views.ts

```diff
@@ -19,82 +19,117 @@
-  const cluster = ClusterStore.getInstance().getById(clusterId);
-
-  if (!cluster || lensViews.has(clusterId)) {
-    return;
+  constructor() {
+    super();
+    makeObservable(this);
+    reaction(() => this.visibleCluster, this.handleVisibleClusterChange);
   }
 
-  logger.info(`[LENS-VIEW]: init dashboard, clusterId=${clusterId}`);
-  const parentElem = document.getElementById("lens-views");
-  const iframe = document.createElement("iframe");
-
-  iframe.id = `cluster-frame-${cluster.id}`;
-  iframe.name = cluster.contextName;
-  iframe.setAttribute("src", getClusterFrameUrl(clusterId));
-  iframe.addEventListener("load", () => {
-    logger.info(`[LENS-VIEW]: loaded from ${iframe.src}`);
-    lensViews.get(clusterId).isLoaded = true;
-  }, { once: true });
-  lensViews.set(clusterId, { clusterId, view: iframe });
-  parentElem.appendChild(iframe);
-
-  logger.info(`[LENS-VIEW]: waiting cluster to be ready, clusterId=${clusterId}`);
-
-  try {
-    await when(() => cluster.ready, { timeout: 5_000 }); // we cannot wait forever because cleanup would be blocked for broken cluster connections
-    logger.info(`[LENS-VIEW]: cluster is ready, clusterId=${clusterId}`);
-  } finally {
-    await autoCleanOnRemove(clusterId, iframe);
+  public hasLoadedView(clusterId: string): boolean {
+    return Boolean(this.views.get(clusterId)?.isLoaded);
   }
-}
 
-export async function autoCleanOnRemove(clusterId: ClusterId, iframe: HTMLIFrameElement) {
-  await when(() => {
+  @action
+  public initView(clusterId: ClusterId) {
     const cluster = ClusterStore.getInstance().getById(clusterId);
 
-    return !cluster || (cluster.disconnected && lensViews.get(clusterId)?.isLoaded);
-  });
-  logger.info(`[LENS-VIEW]: remove dashboard, clusterId=${clusterId}`);
-  lensViews.delete(clusterId);
+    if (!cluster || this.views.has(clusterId)) {
+      return;
+    }
+
+    logger.info(`[LENS-VIEW]: init dashboard, clusterId=${clusterId}`);
+    const parentElem = document.getElementById("lens-views");
+    const iframe = document.createElement("iframe");
+
+    iframe.id = `cluster-frame-${cluster.id}`;
+    iframe.name = cluster.contextName;
+    iframe.style.display = "none";
+    iframe.setAttribute("src", getClusterFrameUrl(clusterId));
+    iframe.addEventListener("load", () => {
+      logger.info(`[LENS-VIEW]: loaded from ${iframe.src}`);
+      this.views.get(clusterId).isLoaded = true;
+    }, { once: true });
+    this.views.set(clusterId, { frame: iframe, isLoaded: false });
+    parentElem.appendChild(iframe);
+
+    logger.info(`[LENS-VIEW]: waiting cluster to be ready, clusterId=${clusterId}`);
+
+    const dispose = when(
+      () => cluster.ready,
+      () => logger.info(`[LENS-VIEW]: cluster is ready, clusterId=${clusterId}`),
+    );
+
+    when(
+      // cluster.disconnect is set to `false` when the cluster starts to connect
+      () => !cluster.disconnected,
+      () => {
+        when(
```

Comment:
these `when()` calls don't need to be disposed?

Reply:
No, because they would only ever be disposed on app quit